cpp/src/arrow/flight/client.cc

```diff
@@ -58,86 +57,81 @@ struct ClientRpc {
+    *out = std::move(message);
+    return Status::OK();
   }
 
- private:
+ protected:
   // The RPC context lifetime must be coupled to the ClientReader
   std::unique_ptr<ClientRpc> rpc_;
-
-  bool stream_finished_;
-  std::shared_ptr<Schema> schema_;
   std::unique_ptr<grpc::ClientReader<pb::FlightData>> stream_;
+  bool stream_finished_;
 };
 
-/// \brief A RecordBatchWriter implementation that writes to a Flight
-/// DoPut stream.
-class FlightPutWriter::FlightPutWriterImpl : public ipc::RecordBatchWriter {
+/// A IpcPayloadWriter implementation that writes to a DoPut stream
+class DoPutPayloadWriter : public ipc::internal::IpcPayloadWriter {
  public:
-  explicit FlightPutWriterImpl(std::unique_ptr<ClientRpc> rpc,
-                               const FlightDescriptor& descriptor,
-                               const std::shared_ptr<Schema>& schema,
-                               MemoryPool* pool = default_memory_pool())
-      : rpc_(std::move(rpc)), descriptor_(descriptor), schema_(schema), pool_(pool) {}
+  DoPutPayloadWriter(const FlightDescriptor& descriptor, std::unique_ptr<ClientRpc> rpc,
+                     std::unique_ptr<protocol::PutResult> response,
+                     std::unique_ptr<grpc::ClientWriter<pb::FlightData>> writer)
+      : descriptor_(descriptor),
+        rpc_(std::move(rpc)),
+        response_(std::move(response)),
+        writer_(std::move(writer)),
+        first_payload_(true) {}
+
+  ~DoPutPayloadWriter() override = default;
+
+  Status Start() override { return Status::OK(); }
 
-  Status WriteRecordBatch(const RecordBatch& batch, bool allow_64bit = false) override {
+  Status WritePayload(const ipc::internal::IpcPayload& ipc_payload) override {
     FlightPayload payload;
-    RETURN_NOT_OK(
-        ipc::internal::GetRecordBatchPayload(batch, pool_, &payload.ipc_message));
+    payload.ipc_message = ipc_payload;
+
+    if (first_payload_) {
+      // First Flight message needs to encore the Flight descriptor
+      DCHECK_EQ(ipc_payload.type, ipc::Message::SCHEMA);
+      std::string str_descr;
+      {
+        pb::FlightDescriptor pb_descr;
+        RETURN_NOT_OK(internal::ToProto(descriptor_, &pb_descr));
+        pb_descr.SerializeToString(&str_descr);
+      }
+      RETURN_NOT_OK(Buffer::FromString(str_descr, &payload.descriptor));
+      first_payload_ = false;
+    }
 
-#ifndef _WIN32
-#pragma GCC diagnostic push
-#pragma GCC diagnostic ignored "-Wstrict-aliasing"
-#endif
-    if (!writer_->Write(*reinterpret_cast<const pb::FlightData*>(&payload),
-                        grpc::WriteOptions())) {
-#ifndef _WIN32
-#pragma GCC diagnostic pop
-#endif
-      std::stringstream ss;
-      ss << "Could not write record batch to stream: "
-         << rpc_->context.debug_error_string();
-      return Status::IOError(ss.str());
+    if (!internal::WritePayload(payload, writer_.get())) {
+      return rpc_->IOError("Could not write record batch to stream: ");
```

Comment:
You can get the underlying gRPC error via `ClientWriter.Finish`: https://grpc.io/grpc/cpp/classgrpc_1_1_client_writer.html#ab6c58e110c289ac0f76ae1fdb8fe24a3

As documented, it will return right away if writing failed (as it would have here): https://grpc.io/grpc/cpp/classgrpc_1_1internal_1_1_client_streaming_interface.html#a47321b8c130947bcef5c793329a54619